src/liballoc/raw_vec.rs

```diff
@@ -561,166 +441,160 @@ impl<T, A: AllocRef> RawVec<T, A> {
+                    .map_err(|_| AllocError { layout, non_exhaustive: () })?
             };
-            let new_layout = Layout::array::<T>(new_cap).map_err(|_| CapacityOverflow)?;
-
-            alloc_guard(new_layout.size())?;
+            memory
+        } else {
+            match placement {
+                MayMove => self.alloc.alloc(layout, init),
+                InPlace => Err(AllocErr),
+            }
+            .map_err(|_| AllocError { layout, non_exhaustive: () })?
+        };
 
-            let res = match self.current_layout() {
-                Some(layout) => {
-                    debug_assert!(new_layout.align() == layout.align());
-                    self.a.realloc(NonNull::from(self.ptr).cast(), layout, new_layout.size())
-                }
-                None => self.a.alloc(new_layout),
-            };
+        self.set_memory(memory);
+        Ok(())
+    }
 
-            let (ptr, new_cap) = match (res, fallibility) {
-                (Err(AllocErr), Infallible) => handle_alloc_error(new_layout),
-                (Err(AllocErr), Fallible) => {
-                    return Err(TryReserveError::AllocError {
-                        layout: new_layout,
-                        non_exhaustive: (),
-                    });
-                }
-                (Ok((ptr, new_size)), _) => (ptr, new_size / elem_size),
-            };
+    fn shrink(
+        &mut self,
+        amount: usize,
+        placement: ReallocPlacement,
+    ) -> Result<(), TryReserveError> {
+        assert!(amount <= self.cap, "Tried to shrink to a larger capacity");
 
-            self.ptr = ptr.cast().into();
-            self.cap = new_cap;
+        let mut memory = if let Some(mem) = self.current_memory() { mem } else { return Ok(()) };
+        let new_size = amount * mem::size_of::<T>();
 
-            Ok(())
+        unsafe {
+            self.alloc.shrink(&mut memory, new_size, placement).map_err(|_| {
+                TryReserveError::AllocError {
+                    layout: Layout::from_size_align_unchecked(new_size, memory.align()),
+                    non_exhaustive: (),
+                }
+            })?;
         }
+
+        self.set_memory(memory);
+        Ok(())
     }
 }
 
 impl<T> RawVec<T, Global> {
-    /// Converts the entire buffer into `Box<[T]>`.
+    /// Converts the entire buffer into `Box<[T]>` with the specified `len`.
     ///
     /// Note that this will correctly reconstitute any `cap` changes
     /// that may have been performed. (See description of type for details.)
     ///
-    /// # Undefined Behavior
+    /// # Safety
     ///
-    /// All elements of `RawVec<T, Global>` must be initialized. Notice that
-    /// the rules around uninitialized boxed values are not finalized yet,
-    /// but until they are, it is advisable to avoid them.
-    pub unsafe fn into_box(self) -> Box<[T]> {
+    /// * `len` must be smaller than or equal to `self.capacity()`
```

Comment:
This safety requirement is not enough: `len` must also "fit" the allocation.

I think a better rule should be:
- `shrink_to_fit(len)` must be called immediately prior to calling this function.

Reply:
Are you sure the rule isn't enough? For ZST `len` can always be `usize::MAX` as a slice with a dangling pointer is created. For sized types, the capacity is equivalent to the number of elements the `RawVec` can hold, so `len` will always fit the allocation. If we didn't allocate at all, the capacity will return `0`.

Reply:
Suppose you have a (non-ZST) `Vec` with a capacity of 100. According to your rule, it is valid for me to pass a `len` of 10 (it is smaller than the capacity). This will fail horribly once the `Box` gets dropped since it will try to free with the wrong size.

Reply:
fair.